### PR TITLE
Add user service

### PR DESCRIPTION
**Pull request by akashmane1598**
## Description
Add user service which saves user data in memory.

### Testing
Please describe the tests that you ran to verify your changes. Please summarize what did you test and what needs to be tested e.g. deployed and tested helm chart locally. 

### Checklist:
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been merged and published in downstream modules

### Documentation
Make sure that you have documented corresponding changes in this repository or [hypertrace docs repo](https://github.com/hypertrace/hypertrace-docs-website) if required.

### CODE REVIEW

File: projects/common/src/user/user.service.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import { HttpClient } from '@angular/common/http';
+import { Injectable } from '@angular/core';
+import { UserTraits } from '../telemetry/telemetry';
+import { InMemoryStorage } from './../utilities/browser/storage/in-memory-storage';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class UserService {
```
Comment (cseas) [inline projects/common/src/user/user.service.ts]: Please rename to UserInfoService or something else. The name `UserService` may easily be confused that it relates to the backend "Hypertrace User Service".

https://hus.concierge.stage.razorpay.in/swagger/index.html

File: projects/common/src/telemetry/providers/rudderstack/rudderstack-provider.ts [hunk 2]
```diff
@@ -27,15 +27,8 @@ export class RudderStackTelemetry implements UserTelemetryProvider<RudderStackCo
...
-    this.http.get<UserTraits>('/user-info').subscribe(
-      (data: UserTraits) => {
-        identify(data.email, { email: data.email, name: data.name });
-      },
-      error => {
-        // tslint:disable-next-line: no-console
-        console.error('something went wrong in identify method', error);
-      }
-    );
+    const data = this.userInfoService.getUserData();
```
Comment (jaywalker21) [inline projects/common/src/telemetry/providers/rudderstack/rudderstack-provider.ts]: Can do this as
`const {email, name} = this.userInfoService.getUserData()`

File: projects/common/src/user/user-info.service.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+@Injectable({
+  providedIn: 'root'
+})
+export class UserInfoService {
+  public BASE_URL: string = '/user-preferences';
+  public static readonly STORAGE_KEY: 'user-data';
+  public static readonly DEFAULT_USER: UserTraits = { name: 'Default User', email: 'ht-user@razorpay.com' };
+  public constructor(private readonly http: HttpClient, private readonly inMemoryStorage: InMemoryStorage) {}
+  public load(): void {
```
Comment (jaywalker21) [inline projects/common/src/user/user-info.service.ts]: Can we return an observable here? The `APP_INITIALIZER` says that we should return a promise or observable to ensure that it is resolved.

File: projects/common/src/user/user-info.service.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+import { UserTraits } from '../telemetry/telemetry';
+import { InMemoryStorage } from '../utilities/browser/storage/in-memory-storage';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class UserInfoService {
+  public BASE_URL: string = '/user-preferences';
+  public static readonly STORAGE_KEY: 'user-data';
+  public static readonly DEFAULT_USER: UserTraits = { name: 'Default User', email: 'ht-user@razorpay.com' };
```
Comment (jaywalker21) [inline projects/common/src/user/user-info.service.ts]: Let's have the `name` as `ht-user` for uniformity?

File: projects/common/src/user/user-info.service.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+  public load(): void {
+    this.http.get<UserTraits>('/user-info').subscribe(
+      (data: UserTraits) => {
+        this.inMemoryStorage.set(UserInfoService.STORAGE_KEY, JSON.stringify(data));
+      },
+      error => {
+        // tslint:disable-next-line: no-console
+        console.error('something went wrong in while fetching user-info', error);
+      }
+    );
```
Comment (jaywalker21) [inline projects/common/src/user/user-info.service.ts]: Can we do a `tap` rather than `subscribe` for the `APP_INITIALISER` part?

**Review by cseas — APPROVED** [review]
LGTM

**Review by jaywalker21 — COMMENTED** [review]
Some comments, if you can go through once.

File: projects/common/src/user/user-info.service.ts [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+  public static readonly DEFAULT_USER: UserTraits = { name: 'Default User', email: 'ht-user@razorpay.com' };
+  public constructor(private readonly http: HttpClient, private readonly inMemoryStorage: InMemoryStorage) {}
+  public load(): void {
+    this.http.get<UserTraits>('/user-info').subscribe(
+      (data: UserTraits) => {
+        this.inMemoryStorage.set(UserInfoService.STORAGE_KEY, JSON.stringify(data));
+      },
+      error => {
+        // tslint:disable-next-line: no-console
+        console.error('something went wrong in while fetching user-info', error);
```
Comment (cseas) [inline projects/common/src/user/user-info.service.ts]: What happens in development mode? Do we always get this error in the console?

File: projects/common/src/user/user-info.service.ts [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+import { catchError, tap } from 'rxjs/operators';
+import { LoggerService } from '../logger/logger.service';
+import { UserTraits } from '../telemetry/telemetry';
+import { InMemoryStorage } from '../utilities/browser/storage/in-memory-storage';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class UserInfoService {
+  public BASE_URL: string = '/user-preferences';
```
Comment (jaywalker21) [inline projects/common/src/user/user-info.service.ts:13]: nit: Not something that needs urgent update, but URL is mostly the whole link, we should have something like `USER_PREFERENCE_ENDPOINT` or something similar here.

**Review by jaywalker21 — APPROVED** [review]
One nit comment, else looks good

**Pull request merged**